### PR TITLE
`azurerm_monitor_diagnostic_setting` - fix update issue 

### PR DESCRIPTION
**Pull request by teowa**
Resolves https://github.com/hashicorp/terraform-provider-azurerm/issues/20019

- set `log_analytics_destination_type` default value to `AzureDiagnostics `  because of service API behaviour change
- fix `hasEnabledLog` logic and correctly set log in API payload when no `log` block or `enabled_log` block has change during resource update. Solve the `at least one type of Log or Metric must be enabled` error.

### CODE REVIEW

File: website/docs/r/monitor_diagnostic_setting.html.markdown [hunk 1]
```diff
@@ -75,25 +74,27 @@ The following arguments are supported:
 
 * `log` - (Optional) One or more `log` blocks as defined below.
 
+-> **NOTE:** `log` has been superseded by `enabled_log` and will be removed in version 4.0 of the AzureRM Provider.
```
Comment (jackofallops) [inline website/docs/r/monitor_diagnostic_setting.html.markdown:77]: We should have a deprecation on the property?

**Review by jackofallops — CHANGES_REQUESTED** [review]
Hi @teowa 
Thanks for this. I think it looks mostly good, but I think there's opportunity for a a couple changes that will help users of this resource. I've left some comments below to take a look at.

Thanks!

File: internal/services/monitor/monitor_diagnostic_setting_resource.go [hunk 2]
```diff
@@ -455,7 +475,7 @@ func resourceMonitorDiagnosticSettingUpdate(d *pluginsdk.ResourceData, meta inte
 		},
 	}
 
-	valid = false
+	valid := false
```
[the remark this answers is not in the document]
Comment (teowa) [reply internal/services/monitor/monitor_diagnostic_setting_resource.go:478]: changed

File: internal/services/monitor/monitor_diagnostic_setting_resource.go [hunk 3]
```diff
@@ -363,13 +364,23 @@ func resourceMonitorDiagnosticSettingUpdate(d *pluginsdk.ResourceData, meta inte
...
 	}
 
+	existing, err := client.Get(ctx, *id)
+	if err != nil {
+		return fmt.Errorf("retrieving Monitor Diagnostics Setting %q for Resource %q: %+v", id.Name, id.ResourceUri, err)
+	}
+	if existing.Model == nil || existing.Model.Properties == nil {
+		return fmt.Errorf("unexpected null model of Monitor Diagnostics Setting %q for Resource %q", id.Name, id.ResourceUri)
+	}
+
```
[the remark this answers is not in the document]
Comment (teowa) [reply internal/services/monitor/monitor_diagnostic_setting_resource.go:374]: changed

File: website/docs/r/monitor_diagnostic_setting.html.markdown (hunk 1 above)
Comment (teowa) [reply website/docs/r/monitor_diagnostic_setting.html.markdown:77]: changed

**Comment by teowa** [comment]
Thanks @jackofallops for reviewing, I have changed the code, please kindly take another look.

**Review by katbyte — APPROVED** [review]
LGTM ⛰️

**Comment by Klaas-** [comment]
I think this introduces a problem, now each time I apply tf wants to set the destination type again:
```
  ~ resource "azurerm_monitor_diagnostic_setting" "diagnostic_settings" {
        id                             = "/subscriptions/[...]"
      + log_analytics_destination_type = "AzureDiagnostics"
        name                           = "LogAnalytics-logs"
        # (2 unchanged attributes hidden)

        # (3 unchanged blocks hidden)
    }
```

**Pull request merged**